Actuator zone modifications and fixes

anilyil opened the pull request:
## Purpose

This PR adds the heat addition feature in the actuator region. Just like adding thrust equally among the actuator zone, we can now also specify an amount of heat power to be added to the flow. This BC variable is differentiated. There is an accompanying PR in baseclasses for the definition of the BC variable. See: https://github.com/mdolab/baseclasses/pull/44 

This also fixes a few more things:
- Bug fix in the differentiation of the flow power integration
- Added a few entries to the gitignore
- Fixed a few docstrings in pyadflow
- Fixed the indentation of the BC printout from Fortran level

I have also created several tests for the actuator zone. These are based on the initial efforts of @joanibal, so thank you Josh! 

The tests include analysis, adjoint, and partial checks. I also added complex tests, and verified the adjoint sensitivities with the complex step. The complex tests will not run right now because they need fixes for the ANK solver. I am creating this PR right now so people can review the changes, and once the complex ANK is also fixed (will do it soon), then we can merge this PR.

## Type of change
What types of change is it?
_Select the appropriate type(s) that describe this PR_

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [x] Documentation update
- [x] Maintenance update
- [ ] Other (please describe)

## Testing

I added new real and complex tests. All tests pass for me locally.

## Checklist

- [ ] I have run `flake8` and `black` to make sure the code adheres to PEP-8 and is consistently formatted
- [x] I have run unit and regression tests which pass locally with my changes
- [x] I have added new tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation

anilyil commented:
I have converted this to a draft because the complex tests won't pass until I fix the complex ANK. Other than that, please take a look and give me feedback on the changes, so we can merge this after the complex solvers are fixed.

anilyil commented:
I just remembered that I also need to add the files required for the tests to the inputfiles tarball. Can you help me with that step @nwu63 @eirikurj ?

anilyil commented:
I have updated the input files, and the baseclasses PR is merged. As expected, the real tests pass now. The complex ones will also pass once we update the complex solvers.

nwu63 commented:
A reminder for @anilyil to bump the minor version here and make a release after this is merged.

anilyil commented:
This one is ready for review @joanibal @sseraj

joanibal commented on tests/reg_tests/test_actuator.py:
```diff
@@ -0,0 +1,814 @@
+        # also add flowpower as an AZ function
+        CFDSolver.addFunction("flowpower", "actuator_region", name="flowpower_az")
+
+    def test_actuator_momentum_flux(self):
```
If it tests both momentum and power we should change the test name

anilyil replied:
So I think I meant by test momentum flux, is to test adding thrust terms, and checking if it has the correct effect on both momentum and energy fluxes through the boundaries (you can add thrust in the actuator zone and it both adds momentum to the flow, and does work on the flow) 

I renamed the tests. Let me know if they are more clear.

joanibal commented on tests/reg_tests/test_actuator.py:
```diff
@@ -0,0 +1,814 @@
+        # just sum these up, the forces contain the correct normals from CFD
+        cfd_force = fp_o + fm_o + fp_i + fm_i
+
+        # the true value is 0, so we add 100 on both sides to have a reasonable comparison
```
This seems like a hack to me. 
Why not just use an absolute tolerance?

anilyil replied:
I don't know how to fix this, and I am probably being dumb but I also could not find a direct answer online. The issue is because the target is exactly zero. Then the relative error ends up being infinite:
```
AssertionError: 
Not equal to tolerance rtol=0.001, atol=0.001

Mismatched elements: 1 / 1 (100%)
Max absolute difference: 0.07964251
Max relative difference: inf
 x: array(0.079643)
 y: array(0.)
```

Do you have any suggestions on how to fix this? I tried setting a rtol of `np.inf` but still fails. I am not sure if there is a version that just checks atol. Swapping the order of value and target also does not help, and that's actually not a great solution in general imo.

What I can do is set the target to the small 0.07964251 value. This is due to the accumulation of integration errors and flow convergence in ADflow. If I compare against this (nonzero) target, then both atol and rtol will work. Or I could add a finite number to both sides (100, or 1 or whatever). Let me know what you think is the best solution here. I am leaving this test in a broken state so we dont merge until this is fixed.

sseraj replied:
I think the `inf` is expected, it's just that the `atol` is too tight.

nwu63 replied:
The function `numpy.testing.assert_allclose` that you used should do exactly what you want, is that not the case? The [docstrings](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html) say it should compare using `atol + rtol * abs(desired)` so if `desired=0` then it should use just `atol`. Maybe you need a looser `atol` since the actual tolerance is around 0.08.

However, if you expect the integration error to be around 0.08 then that may be a better value to use. That will ensure a tighter tolerance and check if in the future this value changes by a small amount.

anilyil replied:
As you correctly pointed out, I am dumb. I set an absolute tolerance of 0.08, and I still compare against a reference of 0. Doing it this was seemed more straightforward for me rather than explaining why the target value is a specific number. I think those type of checks should be done with a comparison to the reference.json file so that they can be trained. 

I think checking with this relatively loose tolerance is okay, because I check all functionals in the test "test_actuator_thrust_and_heat" against the reference values in https://github.com/mdolab/adflow/blob/b3469288f03f99b03032621f01990d836f137513/tests/reg_tests/refs/actuator_tests.json with `rtol=1e-12` and `atol=1e-12`. In summary, I think the tests are good to go imo if you are happy with the final modification.

anilyil replied:
I wanted to keep the comparison to zero just to demonstrate we expect a value of zero here, rather than some other number. Also, I test pretty much everything that touches the actuator zone so I am pretty comfortable with this individual one.

sseraj replied:
I am also fine with this because it basically says that the integration error should not increase past 0.08

joanibal reviewed (changes requested):
A few nitpicks

sseraj commented on tests/reg_tests/test_actuator.py:
```diff
@@ -0,0 +1,814 @@
+        # just sum these up, the forces contain the correct normals from CFD
+        cfd_force = fp_o + fm_o + fp_i + fm_i
+
+        # The low accuracy is because the intgrated quantities don't have a lot of precision
```
"integrated"

anilyil replied:
fixed 3 instances of this typo. words are difficult

anilyil commented on adflow/pyADflow.py:
```diff
@@ -690,9 +724,38 @@ def addActuatorRegion(self, fileName, axis1, axis2, familyName,
+
+        Parameters
+        ----------
+        fileName : str
```
Good catch. I fixed these in `writeActuatorRegions` and `addActuatorRegion`. I also opened an issue linking your comment. Let me know if there are other docstrings I should fix.

nwu63 replied:
Is this true? I haven't found any issues so far. Also, the [style guide](https://numpydoc.readthedocs.io/en/latest/format.html#parameters) explicitly says to have a space before.

We could consider adopting a docstring linter also (pydocstyle comes to mind) in our CI framework. Of course, first step would be to first fix all the violations.

anilyil replied:
I am going to let you two battle it out on this one. let me know if you need me to update this pr.

sseraj replied:
The space before works locally for me but not on RTD. It would be best to follow the style guide though, so maybe we should just try to fix it on the docs build side.

nwu63 replied:
Let's talk offline and figure this out.

sseraj replied:
Fixed by the [latest theme version ](https://github.com/mdolab/sphinx_mdolab_theme/releases/tag/v1.2.1)

anilyil commented on adflow/pyADflow.py:
```diff
@@ -615,13 +615,36 @@ def addIntegrationSurface(self, fileName, familyName, isInflow=True):
+                          thrust=0.0, torque=0.0, heat=0.0,
+                          relaxStart=None, relaxEnd=None):
+        """
+        Add an actuator disk zone defined by the (closed) supplied
```
Yeah good point, I missed this text here. I updated this part of the docstring completely to be more clear.

sseraj replied:
The new docstring is much clearer

anilyil commented on adflow/pyADflow.py:
```diff
@@ -615,13 +615,36 @@ def addIntegrationSurface(self, fileName, familyName, isInflow=True):
+        """
+        Add an actuator disk zone defined by the (closed) supplied
+        in the plot3d file "fileName". axis1 and axis2 defines the
+        physical extent of the region overwhich to apply the ramp
```
Seems like this paragraph was mostly a mess. I updated this part of the docstring completely to be more clear.

sseraj replied:
This typo is still there

anilyil replied:
I should really learn how to read. (I thought you were referring to the not-so-clear explanation there, did not even read that word when I was editing it.) Should be fixed now.

anilyil commented on adflow/pyADflow.py:
```diff
@@ -615,13 +615,36 @@ def addIntegrationSurface(self, fileName, familyName, isInflow=True):
+                          relaxStart=None, relaxEnd=None):
+        """
+        Add an actuator disk zone defined by the (closed) supplied
+        in the plot3d file "fileName". axis1 and axis2 defines the
```
Updated that whole paragraph.

anilyil commented on src/modules/actuatorRegionData.F90:
```diff
@@ -17,13 +17,17 @@ module actuatorRegionData
      real(kind=realType), dimension(3) :: axisVec
+
+     ! Q is the total heat flux to be added on this regoin
+     real(kind=realType) :: Q
```
yeah I agree. it seems we only use these inside the derived `actuatorRegionType`. I was not going to change this but then I realized we also have another %q in flowdoms for viscous surfaces, I am guessing for heat fluxes. I will change these to thrust, torque and heat. We still use them inside the derived data type so having the same parameter name should not affect it in this routine. Worst case, I will rename the inputs with `_in` or sth. I will do this change in a separate commit.

anilyil replied:
See https://github.com/mdolab/adflow/pull/118/commits/b10c7226354e2350787011d5e484406e81eced8f for this fix. Let me know if the new names look ok.

anilyil replied:
See b10c722 for this fix. Let me know if the new names look ok.

sseraj replied:
Thanks, this looks good.

anilyil commented on src/bcdata/BCData.F90:
```diff
@@ -1575,12 +1577,16 @@ subroutine setBCData_d(bcDataNamesIn, bcDataIn, bcDataInd, famLists, sps, &
+                actuatorRegionsd(iRegion)%F = actuatorRegions(iRegion)%axisVec* &
                      bcDataInd(iVar)
              else if (trim(varName) == "Torque") then
+                actuatorRegions(iRegion)%T = bcDataIn(iVar)
```
Yes, setPointer changes are. I fixed this while I was fixing the others. As far as I know, the current torque implementation  does not actually do anything. @shamsheersc19 fixed it at some point for his stuff but that was a separate work. We may look into merging his torque stuff here at some point.

anilyil commented on src/solver/residuals.F90:
```diff
@@ -339,40 +339,47 @@ subroutine sourceTerms_block(nn, res, iRegion, pLocal)
+
+    ! Heat factor. This is heat added per unit volume per unit time
+    Q_fact = factor * actuatorRegions(iRegion)%Q / actuatorRegions(iRegion)%volume / (pRef * uRef * LRef * LRef)
+    mynormsies = (pRef * uRef * LRef * LRef)
```
lol, I forgot to remove this. Probably used it while debugging. removing it.

anilyil commented on src/solver/residuals.F90:
```diff
@@ -315,7 +315,7 @@ subroutine sourceTerms_block(nn, res, iRegion, pLocal)
     use actuatorRegionData
     use blockPointers, only : volRef, dw, w
-    use flowVarRefState, only : Pref, uRef
+    use flowVarRefState, only : pRef, uRef, LRef
```
Because it seems this is defined as `LRef` in https://github.com/mdolab/adflow/blob/master/src/modules/flowVarRefState.F90#L43-L50 

I can change it to `lRef` if you want?

sseraj replied:
That makes sense. `LRef` is fine.

anilyil commented:
thanks for the review @joanibal and @sseraj. I think I addressed all of your comments except Josh's comment at: https://github.com/mdolab/adflow/pull/118#discussion_r599993034 Once we figure out that test, this is good to go for me.

sseraj commented:
I think the only other thing is to bump the minor version

anilyil commented:
I will make the new release after this is merged right?

sseraj commented:
> I will make the new release after this is merged right?

Yes, that would be great.
